google/cloud/aiplatform/base.py

```diff
@@ -561,6 +579,123 @@ def _sync_object_with_future_result(
             if value:
                 setattr(self, attribute, value)
 
+    def _construct_sdk_resource_from_gapic(
+        self,
+        gapic_resource: proto.Message,
+        credentials: Optional[auth_credentials.Credentials] = None,
+    ) -> AiPlatformResourceNoun:
+        """Given a GAPIC object, return the SDK representation."""
+        sdk_resource = self._empty_constructor(credentials=credentials)
+        sdk_resource._gca_resource = gapic_resource
+        return sdk_resource
+
+    # TODO(b/144545165) - Improve documentation for list filtering once available
+    @classmethod
+    def list(
+        cls,
+        filter: Optional[str] = None,
+        order_by: Optional[str] = None,
+        page_size: Optional[int] = None,
+        read_mask: Optional[field_mask.FieldMask] = None,
+    ) -> Sequence[AiPlatformResourceNoun]:
+        """List all instances of this AI Platform Resource.
+
+        Example Usage:
+
+        aiplatform.BatchPredictionJobs.list(
+            filter='state="JOB_STATE_SUCCEEDED" AND display_name="my_job"',
+        )
+
+        aiplatform.Model.list(order_by="create_time desc, display_name")
+
+        Args:
+            filter (str):
+                Optional. An expression for filtering the results of the request.
+                For field names both snake_case and camelCase are supported.
+            order_by (str):
+                Optional. A comma-separated list of fields to order by, sorted in
+                ascending order. Use "desc" after a field name for descending.
+                Supported fields: `display_name`, `create_time`, `update_time`
+            page_size (int):
+                Optional. The standard list page size.
+            read_mask (field_mask.FieldMask):
+                Optional. Mask specifying which fields to read.
+
+        Returns:
+            Sequence[AiPlatformResourceNoun] - A list of SDK resource objects
+        """
+        _UNSUPPORTED_LIST_ORDER_BY_TYPES = (
```

Comment:
Strong preference to break down this conditional logic at the subclass level instead. 

One option is to have two private list classes and pass in the filtering attributes.
```Python
class AiPlatformResourceNoun:
  def _list(..., 
                order_by: Optional[str],
                gapic_field_filter_name: Optional[str]:None,
                cls_field_filter_name: Optional[str]: None):
    ...
    cls_filter_schema = getattr(cls, cls_field_filter_schema, None) if cls_field_filter_name else set([])
    final_list = [
                self._construct_sdk_resource_from_gapic(
                    gapic_resource, credentials=creds
                )
                for gapic_resource in resource_list
                if gapic_field_filter_key and getattr(gapic_resource, gapic_field_filter_key)
                in cls_filter_schema
            ]
  def _list_with_local_order(...,order_by: Optional[str]):
     li = cls._list(..., order_by=None)
     # order here
     return li
     
  def list(...):
    return cls._list(...)
     
class _TrainingJob:
   def list(...):
      return cls._list_with_local_order(
        ...,
        order,
        gapic_field_filter_name='training_task_definition',
        cls_field_filter_name='_supported_training_schemas') # could just pass in the cls attribute as well
 ```

This will be more extensible in the future and allow external teams to use the list functionality without needing to change the base class.

Reply:
Thanks for putting together this example! Adopting that option but will have the private `_list` method take a single `cls_filter` Callable that takes a `gca_resource` returns a bool that decides whether to include or exclude a given GAPIC object. Lmk if you have any concerns with that approach.